hydromt_wflow/wflow.py

```diff
@@ -2715,6 +2716,267 @@ def setup_precip_forcing(
+        if (
+            isinstance(precip_fn, pd.DataFrame)
+            or isfile(precip_fn)
+            or _data_type == "DataFrame"
+        ):
+            df_precip = self.data_catalog.get_dataframe(
+                precip_fn,
+                time_tuple=(starttime, endtime),
+                variables=station_ids,
+                handle_nodata=NoDataStrategy.IGNORE,
+            )
+            if precip_stations_fn is None and interp_type != "uniform":
+                raise ValueError(
+                    """Using a DataFrame as precipitation source requires that \
+station locations are provided separately through precip_station_fn."""
+                )
+            # For uniform: switch to nearest-neighbour which works with single station
+            if interp_type == "uniform":
+                if df_precip.shape[1] != 1:
+                    raise ValueError(
+                        f"""
+                        Data source ({precip_fn}) should contain
+                        a single timeseries, not {df_precip.shape[1]}."""
+                    )
+                interp_type = "nearest"
+                self.logger.info(
+                    "Uniform interpolation is applied using method 'nearest'.")
+            # For other methods: raise error when a DataFrame is loaded without stations
+            elif precip_stations_fn is None:
+                raise ValueError(
+                    """Using a DataFrame as precipitation source requires that \
+station locations are provided separately through precip_station_fn."""
+                )
+
+        # Load precip as GeoDataset, which does not require precip_stations_fn
+        elif isinstance(precip_fn, xr.Dataset) or _data_type == "GeoDataset":
+            # TODO:
+            # da_precip = self.data_catalog.get_geodataset(...) using geom_buffered
+            # df_precip = pd.DataFrame(...)
+            # precip_stations_fn = gpd.GeoDataFrame(...)
+            raise NotImplementedError("GeoDataset source not yet implented.")
+        else:
+            raise ValueError(f"Data source {precip_fn} not recognized.")
+
+        df_precip = df_precip.astype("float32")
+
+        # Check number of stations, at least 3 needed to interpolate spatially
+        if (len(df_precip.columns) < 3) & (interp_type != "nearest"):
+            raise ValueError(
+                f"""Only {len(df_precip.columns)} found with timeseries, which is \
+not sufficient for interpolation using {interp_type} (3 or more required)."""
+            )
+
+        # Transform station coordinates to model crs
+        gdf_stations = gdf_stations.to_crs(self.crs)
+
+        # Use model resolution for the interpolated grid
+        if not np.allclose((np.abs(self.res)), 1e-6):
+            self.logger.info(
+                f"""Mismatch in horizontal and vertical resolution ({self.res}), \
+using finest resolution during interpolation by MetPy."""
+            )
```

Comment:
This info can be moved to interp function. This will avoid repeat if we want to implement for temperature/pet or when we do support geodataset.